core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -821,11 +858,60 @@ object Parser {
 
   /** Represents where a failure occurred and all the expectations that were broken
     */
-  final case class Error(failedAtOffset: Int, expected: NonEmptyList[Expectation]) {
+  final case class Error(input: String, failedAtOffset: Int, expected: NonEmptyList[Expectation]) {
     def offsets: NonEmptyList[Int] =
       expected.map(_.offset).distinct
   }
 
+  object Error {
+    implicit val catsShowError: Show[Error] =
+      new Show[Error] {
+        def show(error: Error): String = {
+          val locationMap = new LocationMap(error.input)
+          val errorMsg = error.expected.map(e => s"* ${e.show}").toList.mkString("\n")
+
+          locationMap.toCaret(error.failedAtOffset).fold(errorMsg) { caret =>
+            val lines = error.input.split('\n')
```

Comment:
the `LocationMap` has the lines and you can get it from there. We shouldn't split it again.